{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"nerve.ipynb","provenance":[],"collapsed_sections":[]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"eLc1ciFLp02_"},"source":["# NERVE 2"]},{"cell_type":"markdown","metadata":{"id":"xOYlvyAnp3Ub"},"source":["## Intro: install all the dependencies/packets/modules"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Ob3lAcGOBexU","executionInfo":{"status":"ok","timestamp":1631821393149,"user_tz":-120,"elapsed":99495,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}},"outputId":"0ae87155-5149-43dd-d8fc-9320a80dda94"},"source":["!rm -r spaan\n","!rm -r NERVE\n","!git clone git://github.com/nicolagulmini/spaan\n","!git clone git://github.com/nicolagulmini/NERVE\n","!python -m pip install git+https://github.com/nicolagulmini/tmhmm.py\n","!pip install Bio\n","!pip install Pandas\n","!apt-get install ncbi-blast+ # for the autoimmunity module if we meant to use blastp to make the comparisons"],"execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["rm: cannot remove 'spaan': No such file or directory\n","rm: cannot remove 'NERVE': No such file or directory\n","Cloning into 'spaan'...\n","remote: Enumerating objects: 195, done.\u001b[K\n","remote: Counting objects: 100% (195/195), done.\u001b[K\n","remote: Compressing objects: 100% (155/155), done.\u001b[K\n","remote: Total 195 (delta 109), reused 91 (delta 37), pack-reused 0\u001b[K\n","Receiving objects: 100% (195/195), 5.63 MiB | 28.14 MiB/s, done.\n","Resolving deltas: 100% (109/109), done.\n","Cloning into 'NERVE'...\n","remote: Enumerating objects: 117, done.\u001b[K\n","remote: Counting objects: 100% (117/117), done.\u001b[K\n","remote: Compressing objects: 100% (102/102), done.\u001b[K\n","remote: Total 117 (delta 52), reused 51 (delta 13), pack-reused 0\u001b[K\n","Receiving objects: 100% (117/117), 17.24 MiB | 18.27 MiB/s, done.\n","Resolving deltas: 100% (52/52), done.\n","Collecting git+https://github.com/nicolagulmini/tmhmm.py\n","  Cloning https://github.com/nicolagulmini/tmhmm.py to /tmp/pip-req-build-x_ipeuxl\n","  Running command git clone -q https://github.com/nicolagulmini/tmhmm.py /tmp/pip-req-build-x_ipeuxl\n","Requirement already satisfied: numpy>=1.9 in /usr/local/lib/python3.7/dist-packages (from tmhmm.py==1.3.1) (1.19.5)\n","Building wheels for collected packages: tmhmm.py\n","  Building wheel for tmhmm.py (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for tmhmm.py: filename=tmhmm.py-1.3.1-cp37-cp37m-linux_x86_64.whl size=160103 sha256=bc960e4d91cd3f5c329e9179fb1a1abeae46d3209da94c50a90ec785379f5675\n","  Stored in directory: /tmp/pip-ephem-wheel-cache-sghtn29t/wheels/fe/53/f7/6b6ddefd0b816f60565cde13047c58f306cba76c21e9035db6\n","Successfully built tmhmm.py\n","Installing collected packages: tmhmm.py\n","Successfully installed tmhmm.py-1.3.1\n","Collecting Bio\n","  Downloading bio-0.8.9-py3-none-any.whl (84 kB)\n","\u001b[K     |████████████████████████████████| 84 kB 2.3 MB/s \n","\u001b[?25hCollecting biopython>=1.79\n","  Downloading biopython-1.79-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (2.3 MB)\n","\u001b[K     |████████████████████████████████| 2.3 MB 9.0 MB/s \n","\u001b[?25hCollecting mygene\n","  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n","Requirement already satisfied: tqdm>=4.62 in /usr/local/lib/python3.7/dist-packages (from Bio) (4.62.0)\n","Requirement already satisfied: plac in /usr/local/lib/python3.7/dist-packages (from Bio) (1.1.3)\n","Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from Bio) (2.23.0)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython>=1.79->Bio) (1.19.5)\n","Collecting biothings-client>=0.2.6\n","  Downloading biothings_client-0.2.6-py2.py3-none-any.whl (37 kB)\n","Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (1.24.3)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2021.5.30)\n","Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2.10)\n","Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (3.0.4)\n","Installing collected packages: biothings-client, mygene, biopython, Bio\n","Successfully installed Bio-0.8.9 biopython-1.79 biothings-client-0.2.6 mygene-3.2.2\n","Requirement already satisfied: Pandas in /usr/local/lib/python3.7/dist-packages (1.1.5)\n","Requirement already satisfied: numpy>=1.15.4 in /usr/local/lib/python3.7/dist-packages (from Pandas) (1.19.5)\n","Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.7/dist-packages (from Pandas) (2.8.2)\n","Requirement already satisfied: pytz>=2017.2 in /usr/local/lib/python3.7/dist-packages (from Pandas) (2018.9)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil>=2.7.3->Pandas) (1.15.0)\n","Reading package lists... Done\n","Building dependency tree       \n","Reading state information... Done\n","The following additional packages will be installed:\n","  ncbi-data\n","The following NEW packages will be installed:\n","  ncbi-blast+ ncbi-data\n","0 upgraded, 2 newly installed, 0 to remove and 40 not upgraded.\n","Need to get 13.1 MB of archives.\n","After this operation, 66.7 MB of additional disk space will be used.\n","Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 ncbi-data all 6.1.20170106-2 [3,645 kB]\n","Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 ncbi-blast+ amd64 2.6.0-1 [9,446 kB]\n","Fetched 13.1 MB in 1min 9s (189 kB/s)\n","Selecting previously unselected package ncbi-data.\n","(Reading database ... 148492 files and directories currently installed.)\n","Preparing to unpack .../ncbi-data_6.1.20170106-2_all.deb ...\n","Unpacking ncbi-data (6.1.20170106-2) ...\n","Selecting previously unselected package ncbi-blast+.\n","Preparing to unpack .../ncbi-blast+_2.6.0-1_amd64.deb ...\n","Unpacking ncbi-blast+ (2.6.0-1) ...\n","Setting up ncbi-data (6.1.20170106-2) ...\n","Setting up ncbi-blast+ (2.6.0-1) ...\n","Processing triggers for hicolor-icon-theme (0.17-2) ...\n","Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n"]}]},{"cell_type":"code","metadata":{"id":"q-tB0jg5B-zw","executionInfo":{"status":"ok","timestamp":1631821395016,"user_tz":-120,"elapsed":1877,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}}},"source":["import tmhmm                                                # to predict transmembrane domains\n","from Bio import SeqIO                                       # to handle .fasta files\n","from Bio.Blast.Applications import NcbiblastpCommandline    # autoimmunity module, to query the local sapiens database\n","from Bio.ExPASy.ScanProsite import scan, read               # scan prosite functional domains information\n","from Bio.Blast import NCBIXML                               # to parse the peptides\n","from NERVE import Protein                                   # to contain proteins information\n","from tensorflow import keras                                # to use the spaan model (to predict the probability of a protein to be an adhesin)\n","import pandas                                               # to read mhcpep.csv and write the final report\n","from spaan.data_processing import *                     # to extract proteins features for the spaan model"],"execution_count":2,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"gj1u_5NQqdFx"},"source":["## User parameters"]},{"cell_type":"code","metadata":{"id":"QnDaOM2tCLrs","executionInfo":{"status":"ok","timestamp":1631821395017,"user_tz":-120,"elapsed":11,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}}},"source":["p_ad_no_citoplasm_filter = 0.46\n","p_ad_extracellular_filter = 0.38\n","transmemb_dom_limit_filter = 2\n","e_value = 1e-10\n","similarity_function = 0.8\n","verbose = 0\n","GRAM = 'p' # for now we can mantain it 'p' as 'positive', then the user will have to put the right parameter! p = positive, n = negative, a = archea (for psortb)"],"execution_count":3,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"cR_1eti6qfai"},"source":["## Define all the initial objects/paths"]},{"cell_type":"code","metadata":{"id":"tVXFZ77cCQvm","executionInfo":{"status":"ok","timestamp":1631821395017,"user_tz":-120,"elapsed":10,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}}},"source":["# main program\n","\n","# TODO: check parameters, license ecc.\n","\n","path_to_fastas = \"./NERVE/data/tre.fasta\"\n","path_to_another_proteome = \"./NERVE/data/tmp.fasta\"\n","list_of_fasta_proteins = list(SeqIO.parse(path_to_fastas, \"fasta\")) # put the right path\n","\n","list_of_proteins = []\n","for p in list_of_fasta_proteins:\n","\tp_id = p.id\n","\tp_seq = p.seq\n","\tlist_of_proteins.append(Protein.Protein(p_id, p_seq))\n","\n","if verbose > 0:\n","\tfor p in list_of_proteins:\n","\t\tp.print_information()"],"execution_count":4,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"BDN_TkUTqlyP"},"source":["# Subcelloc "]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"id":"a8XzKlsgHvYb","executionInfo":{"status":"ok","timestamp":1631821395025,"user_tz":-120,"elapsed":18,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}},"outputId":"e6668dbf-5f23-44d9-9e57-1be433301ac5"},"source":["# psortb test\n","'''\n","6. Limitations\n","    6.1 Proteins resident at multiple localization sites: Many proteins can exist at multiple localization sites. Examples of such proteins include integral membrane proteins with large periplasmic domains, or autotransporters, which contain an outer membrane pore domain and a cleaved extracellular domain. The current version of PSORTb handles this situation by flagging proteins which show a distribution of localization scores favouring two sites, rather than one. It is important to examine the distribution of localization scores carefully in order to determine if your submitted protein may have multiple localization sites and if so, which two sites are involved.\n","    6.2 Lipoproteins: The current version of PSORTb does not detect lipoprotein motifs.\n","    6.3 Precision vs. Recall: PSORTb is designed to emphasize precision (or specificity) over recall (or sensitivity). Programs which make predictions at all costs often provide incorrect or incomplete results, which can be propagated through annotated databases, datasets and reports in the literature. We believe that a confident prediction is more valuable than any prediction, and we have designed the program to this end. Note, however, that a user may choose to use their own reduced cutoff score in generating final predictions.\n","'''\n","# use the psortb.py in python code that I found online (link in the comment of the file itself). Then try it with linux."],"execution_count":5,"outputs":[{"output_type":"execute_result","data":{"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"},"text/plain":["'\\n6. Limitations\\n    6.1 Proteins resident at multiple localization sites: Many proteins can exist at multiple localization sites. Examples of such proteins include integral membrane proteins with large periplasmic domains, or autotransporters, which contain an outer membrane pore domain and a cleaved extracellular domain. The current version of PSORTb handles this situation by flagging proteins which show a distribution of localization scores favouring two sites, rather than one. It is important to examine the distribution of localization scores carefully in order to determine if your submitted protein may have multiple localization sites and if so, which two sites are involved.\\n    6.2 Lipoproteins: The current version of PSORTb does not detect lipoprotein motifs.\\n    6.3 Precision vs. Recall: PSORTb is designed to emphasize precision (or specificity) over recall (or sensitivity). Programs which make predictions at all costs often provide incorrect or incomplete results, which can be propagated through annotated databases, datasets and reports in the literature. We believe that a confident prediction is more valuable than any prediction, and we have designed the program to this end. Note, however, that a user may choose to use their own reduced cutoff score in generating final predictions.\\n'"]},"metadata":{},"execution_count":5}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"1v9n_Pw0CeY3","executionInfo":{"status":"ok","timestamp":1631821395026,"user_tz":-120,"elapsed":14,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}},"outputId":"24c1e5a5-2fc2-4c03-db29-0615d82cc8d9"},"source":["# Subcelloc module\n","# ... cerca di capire come chiamare psortb da qua\n","print(\"Warning: the PSORTB output has to be 'terse'!\")\n","\n","# final predictions parsing from the output of psortb program\n","filename = \"20210903075318_psortb_gramneg.txt\" # put the right filename from the previous (and still absent) lines of code\n","fp = open(\"./NERVE/data/\"+filename, 'r')\n","lines = fp.readlines()\n","for i in range(1, len(lines)):\n","    attributes = lines[i].split(' ')\n","    if attributes[0] == list_of_proteins[i-1].id:\n","        last_attributes = attributes[len(attributes)-1].split('\\t') # example: ['SV=2', 'Cytoplasmic', '9.97\\n']\n","        list_of_proteins[i-1].localization = last_attributes[1] # if it gives errors, try last_attributes[len(last_attributes)-2]\n","fp.close()"],"execution_count":6,"outputs":[{"output_type":"stream","name":"stdout","text":["Warning: the PSORTB output has to be 'terse'!\n"]}]},{"cell_type":"markdown","metadata":{"id":"OwGDvGM3qni_"},"source":["# Adhesin"]},{"cell_type":"code","metadata":{"id":"xu2-nw1kCewm","executionInfo":{"status":"ok","timestamp":1631821396755,"user_tz":-120,"elapsed":1742,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}}},"source":["# Adhesin module\n","\n","model = keras.models.load_model('./NERVE/espaan_model.h5') # insert the path to location!\n","for p in list_of_proteins:\n","\tp.p_ad = float(model.predict([\n","\t\t\t     np.array([aminoacids_frequencies(p.sequence)]),\n","\t\t\t     np.array([multiplet_frequencies(p.sequence, 3)]),\n","\t\t\t     np.array([multiplet_frequencies(p.sequence, 4)]),\n","\t\t\t     np.array([multiplet_frequencies(p.sequence, 5)]),\n","\t\t\t     np.array([dipeptide_frequencies(p.sequence)]),\n","\t\t\t     np.array([charge_composition(p.sequence)]),\n","\t\t\t     np.array([hydrophobic_composition(p.sequence)])\n","\t\t     ]))"],"execution_count":7,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"kQbSXC2Fqo2P"},"source":["# Tmhelices"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"SOnJbnpVEaac","executionInfo":{"status":"ok","timestamp":1631821397034,"user_tz":-120,"elapsed":281,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}},"outputId":"6578a4a6-1a25-45c0-ad51-961409dc7f8c"},"source":["# Tmhelices module\n","\n","for p in list_of_proteins:\n","    annotation, _ = tmhmm.predict(p.sequence)\n","    p.tmhmm_seq = annotation\n","    transmembrane_domains = 0\n","    for i in range(len(annotation)-1):\n","        if (annotation[i] == 'i' or annotation[i] == 'o') and annotation[i+1] == 'M':\n","            transmembrane_domains += 1\n","    p.transmembrane_doms = transmembrane_domains\n","for p in list_of_proteins:\n","    if p.transmembrane_doms > 2:\n","        p.transmembrane_flag = 1"],"execution_count":8,"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/tmhmm/api.py:20: RuntimeWarning: divide by zero encountered in log\n","  _, path = viterbi(sequence, *model)\n"]}]},{"cell_type":"markdown","metadata":{"id":"SJ0UDXjTqql9"},"source":["# Autoimmunity"]},{"cell_type":"code","metadata":{"id":"DVpGqY1Vx0tz","executionInfo":{"status":"ok","timestamp":1631821397582,"user_tz":-120,"elapsed":557,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}}},"source":["# Autoimmunity module\n","\n","blastx_cline = NcbiblastpCommandline(query=path_to_fastas, db=\"./NERVE/sapiens_database/sapiens\", evalue=e_value, outfmt=5, out=\"./sapiens.xml\") # 5 is for xml \n","stdout, stderr = blastx_cline()\n","# print the warning: \"you can find a sapiens.xml file on your working directory which is the outputs of the autoimmunity module. Do not delete during the computation. After the computation it will be deleted in order to avoid future collisions.\"\n","# poi cancella l'output"],"execution_count":9,"outputs":[]},{"cell_type":"code","metadata":{"id":"GuAIEnYuDrIE","executionInfo":{"status":"ok","timestamp":1631821397583,"user_tz":-120,"elapsed":4,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}}},"source":["# for each result in the .xml file...\n","for record in NCBIXML.parse(open(\"./sapiens.xml\")):\n","\n","    query_name = record.query.split(' ')[0] # take only the query id \n","    #print(\"QUERY: \" + record.query)\n","\n","    # take the right candidate to update\n","    tmp_protein = list_of_proteins[0]\n","    for p in list_of_proteins:\n","        if p.id == query_name:\n","            tmp_protein = p\n","\n","    #print(\"Check the id of the corresponding candidate: \" + tmp_protein.id)\n","    # for each effective alignment between the tmp candidate and the human proteome\n","    for alignment in record.alignments:\n","        # collect all the interesting peptides\n","        for hsp in alignment.hsps:\n","            tmp_protein.list_of_shared_human_peps += Protein.Protein.hsp_match_parser(hsp.match)\n","\n","    # print out the peptides (if there are any)\n","    '''\n","    if len(tmp_protein.list_of_shared_human_peps) == 0:\n","        print(\"No interesting peptides.\")\n","    else:\n","        print(\"List of interesting peptides: \" + str(tmp_protein.list_of_shared_human_peps))\n","    print()\n","    '''"],"execution_count":10,"outputs":[]},{"cell_type":"code","metadata":{"id":"R2_2FJ4T1dE1","executionInfo":{"status":"ok","timestamp":1631821397955,"user_tz":-120,"elapsed":375,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}}},"source":["# mhcpep pandas visualization\n","\n","mhcpep = pandas.read_csv(\"./NERVE/mhcpep.csv\")\n","# quali sono i peptidi?\n","# come opero il confronto?\n","# realizzare poi la somma della formula "],"execution_count":11,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ug53JaHzpu8T"},"source":["# Conservation (optional)"]},{"cell_type":"code","metadata":{"id":"0skh4FwT9U5b","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1631821397955,"user_tz":-120,"elapsed":12,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}},"outputId":"efe34245-0fff-485c-d921-3c3220ba0990"},"source":["# (optional) Conservation module \n","# mettere tutto in un if path_to_another_proteome is not None\n","#!declare -r path_to_another_proteome = \"./NERVE/data/tmp.fasta\"\n","# capire come definire una variable in shell\n","!makeblastdb -in \"./NERVE/data/tre.fasta\" -dbtype prot -parse_seqids -out \"./compare_proteome/compare_proteome\""],"execution_count":12,"outputs":[{"output_type":"stream","name":"stdout","text":["\n","\n","Building a new DB, current time: 09/16/2021 19:43:16\n","New DB name:   /content/compare_proteome/compare_proteome\n","New DB title:  ./NERVE/data/tre.fasta\n","Sequence type: Protein\n","Keep MBits: T\n","Maximum file size: 1000000000B\n","Adding sequences from FASTA; added 3 sequences in 0.000518799 seconds.\n"]}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"vsUH9201EORm","executionInfo":{"status":"ok","timestamp":1631821440591,"user_tz":-120,"elapsed":228,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}},"outputId":"d58a910f-e54c-421e-8f3a-ef259b7e1065"},"source":["blastx_cline = NcbiblastpCommandline(query=path_to_fastas, db=\"./compare_proteome/compare_proteome\", evalue=e_value, outfmt=5, out=\"./comparison.xml\") # 5 is for xml \n","stdout, stderr = blastx_cline()\n","\n","for record in NCBIXML.parse(open(\"./comparison.xml\")):\n","    query_name = record.query.split(' ')[0] \n","\n","    tmp_protein = list_of_proteins[0]\n","    for p in list_of_proteins:\n","        if p.id == query_name:\n","            tmp_protein = p\n","    # max score between all the alignments\n","    max_score = 0\n","    for alignment in record.alignments:\n","        # collect all the interesting peptides\n","        for hsp in alignment.hsps:\n","            if hsp.score > max_score:\n","                max_score = hsp.score\n","    \n","    tmp_protein.conservation_score  = max_score"],"execution_count":14,"outputs":[{"output_type":"stream","name":"stdout","text":["QUERY: sp|P06846|EBGR_ECOLI HTH-type transcriptional regulator EbgR OS=Escherichia coli (strain K12) OX=83333 GN=ebgR PE=4 SV=2\n","QUERY: sp|P0AAC4|YBHL_ECOLI Inner membrane protein YbhL OS=Escherichia coli (strain K12) OX=83333 GN=ybhL PE=1 SV=1\n","QUERY: sp|P0AE16|AMPG_ECOLI Anhydromuropeptide permease OS=Escherichia coli (strain K12) OX=83333 GN=ampG PE=1 SV=1\n"]}]},{"cell_type":"markdown","metadata":{"id":"ju5wc-iXqtNe"},"source":["# Function"]},{"cell_type":"code","metadata":{"id":"-AbunEdA9uUp","executionInfo":{"status":"aborted","timestamp":1631821398507,"user_tz":-120,"elapsed":327,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}}},"source":["# Function module\n","# (make blastp on uniprot... NOOOOOO) #!makeblastdb -in \"./uniprot_sprot.fasta\" -dbtype prot -parse_seqids -out \"uniprot\"\n","\n","# prosite information\n","for p in list_of_proteins:\n","    handler = scan(p.sequence)\n","    p.scan_prosite_information = read(scan(p.sequence))\n","\n","\n","# try deepgo! https://deepgo.cbrc.kaust.edu.sa/deepgo/"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"QgjggufEqux_"},"source":["# Select"]},{"cell_type":"code","metadata":{"id":"LlsKlxF89xso","executionInfo":{"status":"aborted","timestamp":1631821398507,"user_tz":-120,"elapsed":5,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}}},"source":["# Select module (filters)\n","# define the ranking"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Nzpoz00HqwKR"},"source":["## Print out all the retrieved information about each candidate of the given proteome"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"I3IzoiupUPvZ","executionInfo":{"status":"ok","timestamp":1631821466148,"user_tz":-120,"elapsed":214,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}},"outputId":"f676d67b-73ab-4854-e2f8-72840399db33"},"source":["for protein in list_of_proteins:\n","    protein.print_information()"],"execution_count":15,"outputs":[{"output_type":"stream","name":"stdout","text":["Information about protein sp|P06846|EBGR_ECOLI:\n","   accession number = P06846\n","   length = 327\n","   localization = Cytoplasmic\n","   estimated probability to be an adhesin = 0.01151728630065918\n","   number of transmembrane domains = 0\n","   no interesting peptides shared with sapiens\n","   conservation with the given proteome blastp score = 1730.0\n","   putative function of the protein = not yet implemented\n","\n","Information about protein sp|P0AAC4|YBHL_ECOLI:\n","   accession number = P0AAC4\n","   length = 234\n","   localization = CytoplasmicMembrane\n","   estimated probability to be an adhesin = 0.04193472862243652\n","   number of transmembrane domains = 7\n","   list of interesting peptides shared with sapiens = ['+L A++LYL', 'L A++LYLD', 'A++LYLD I', '++LYLD IN', '+LYLD INL', 'LYLD INLF', 'YLD INLFL', 'INLFL LLR']\n","   conservation with the given proteome blastp score = 1186.0\n","   putative function of the protein = not yet implemented\n","\n","Information about protein sp|P0AE16|AMPG_ECOLI:\n","   accession number = P0AE16\n","   length = 491\n","   localization = CytoplasmicMembrane\n","   estimated probability to be an adhesin = 0.02877509593963623\n","   number of transmembrane domains = 14\n","   no interesting peptides shared with sapiens\n","   conservation with the given proteome blastp score = 2507.0\n","   putative function of the protein = not yet implemented\n","\n"]}]},{"cell_type":"code","metadata":{"id":"mbtGkxTrUVYl","executionInfo":{"status":"aborted","timestamp":1631821398510,"user_tz":-120,"elapsed":7,"user":{"displayName":"Nicola Gulmini","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiRSJO1pD88AsDH3UdbhNdHDtzum41rsWGwWB_J6w=s64","userId":"02235667290947110441"}}},"source":[""],"execution_count":null,"outputs":[]}]}